thelounge: init at 3.0.1

Mrmaxmeier opened the pull request:
###### Motivation for this change

[The Lounge](https://github.com/thelounge/thelounge) is the official and community-managed fork of Shout - _The self-hosted web IRC client_.
The main motivation for inclusion to nixpkgs is that I don't really know how to maintain packages out-of-tree `(:`

###### Things done
- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

I've been running this for a few days.

A few things to consider:

- The service should be compatible with the soon-to-be-released `v3.0` of The Lounge. The update to `v3` might happen during unrelated rebuilds of `nodePackages` because the version is not pinned. Not sure if that's ideal. It's in line with `shout` package though.
- ~User-management for private instances is not (yet?) handled by this PR. It might be useful to include a script that sets `THELOUNGE_HOME` and invokes the `thelounge` binary as the `thelounge` user.~
- The `shout` package and service should probably be marked as deprecated or for removal. Is there a standard way to do this?
- There might be a more canonical way to handle mostly-known config options.

Mic92 commented on nixos/modules/services/networking/thelounge.nix:
```diff
@@ -0,0 +1,85 @@
+
+    home = mkOption {
+      type = types.path;
+      default = "/var/lib/thelounge";
```
I don't think there is a strong need for this to be configurable.
This can be manged using systemd's  `StateDirectory=` that take also care of changing the permission on service startup.

Mic92 commented on nixos/modules/misc/ids.nix:
```diff
@@ -338,6 +338,7 @@
       minetest = 311;
       rss2email = 312;
       cockroachdb = 313;
+      thelounge = 314;
```
As this service only will allocate a few files, then become dynamic ids as well.

Mic92 commented on nixos/modules/services/networking/thelounge.nix:
```diff
@@ -0,0 +1,85 @@
+      description = "The Lounge daemon user";
+      home = cfg.home;
+      createHome = true;
+    };
```
When you use systemd's `StateDirectory` you can remove home and createHome frome here.

Mic92 commented on nixos/modules/services/networking/thelounge.nix:
```diff
@@ -0,0 +1,85 @@
+      description = "The Lounge web IRC client";
+      wantedBy = [ "multi-user.target" ];
+      wants = [ "network-online.target" ];
+      after = [ "network-online.target" ];
```
I don't think this service needs `network-online.target`, does it?
Would it not just try to reconnect after some time?

Mic92 replied:
It is delaying the service start for no good reason.

Mic92 replied:
You can also checkout the last section here: https://www.freedesktop.org/wiki/Software/systemd/NetworkTarget/

Mic92 commented:
When removing should, you can add a entry to our 19.03 release notes.

Mic92 commented on nixos/modules/services/networking/thelounge.nix:
```diff
@@ -0,0 +1,79 @@
+  );
+
+  thelounge-wrapper-script = pkgs.writeShellScriptBin "thelounge" ''
+    systemd-run --pty --property=DynamicUser=yes \
```
Would a systemd user service and why should it write to `/var/lib/thelounge` instead of the users home?

Mic92 replied:
Actually this is a bit out-of-scope of nixos and could be part of home-manager instead.

Mic92 replied:
now I got the idea, this is basically used to add new users. But would that actually run the same dynamic user that thelounge was started with?

Mrmaxmeier replied:
Yup. User-management is done with `thelounge add` etc. This becomes really awkward with systemd's DynamicUser because sessions using `sudo -u thelounge` somehow can't access `/var/lib/thelounge`.

Not sure whether `systemd-run` is the preferred solution. It doesn't seem like there's any other package in the same situation.

Mic92 replied:
This will not work. There is a new uid on every run.

```console
$ sudo systemd-run --pty --property=DynamicUser=yes id
Running as unit: run-u1538.service
Press ^] three times within 1s to disconnect TTY.
uid=62575 gid=62575 groups=62575
$ sudo systemd-run --pty --property=DynamicUser=yes id
Running as unit: run-u1537.service
Press ^] three times within 1s to disconnect TTY.
uid=61729 gid=61729 groups=61729
```

I would suggest to create a user using `users.users` without allocating a static uid. When the module is configured for the first time, this will assign a uid and a passwd entry so that sudo can be used.

Mrmaxmeier replied:
You're missing `--property=StateDirectory=thelounge`.
```
$ sudo systemd-run --pty --property=DynamicUser=yes --property=StateDirectory=thelounge id
Running as unit: run-u751.service
Press ^] three times within 1s to disconnect TTY.
uid=64374 gid=64374 groups=64374
$ sudo systemd-run --pty --property=DynamicUser=yes --property=StateDirectory=thelounge id
Running as unit: run-u755.service
Press ^] three times within 1s to disconnect TTY.
uid=64374 gid=64374 groups=64374
```

Infinisil replied:
I don't think this is a good solution. `DynamicUser` works well when the service can run truly isolated, but in this case the user needs to interact with it, so it's not isolated anymore, and this is a workaround for that.

Instead you can just define a new user for thelounge (no static uid needed though) and use that instead of DynamicUser.

nornagon commented:
Is there anything I can do to help out getting this landed?

(And/or, how do I patch this in locally to test it out...?)

Infinisil commented on nixos/modules/services/networking/thelounge.nix:
```diff
@@ -0,0 +1,77 @@
+        The options defined here will be merged to the default configuration file.
+        Note: In case of duplicate configuration, options from extraConfig have priority.
+
+        Documentation: https://thelounge.chat/docs/server/configuration
```
You need something like `<link xlink:href="https://example.com"/>`

Infinisil commented on nixos/modules/services/networking/thelounge.nix:
```diff
@@ -0,0 +1,77 @@
+        converted to JSON to generate the configuration file).
+
+        The options defined here will be merged to the default configuration file.
+        Note: In case of duplicate configuration, options from extraConfig have priority.
```
`<option>extraConfig</option>`

Infinisil commented on nixos/modules/services/networking/thelounge.nix:
```diff
@@ -0,0 +1,77 @@
+      default = {};
+      type = types.attrs;
+      example = {
+        reverseProxy = true;
```
Needs
```nix
{
  example = literalExample ''{
    # ...
  }'';
}
```

Infinisil commented on nixos/modules/services/networking/thelounge.nix:
```diff
@@ -0,0 +1,77 @@
+
+    extraConfig = mkOption {
+      default = {};
+      type = types.attrs;
```
This would be a potential candidate for NixOS/rfcs#42, but that's unfortunately not yet accepted. Having the type as `types.attrs` here makes merging fail for nested values.

Mrmaxmeier replied:
I'll consider this once the RFC is accepted. I don't think this affects `thelounge` a bunch though because the surface area of configuration managed in the service definition is pretty small.

Infinisil commented on nixos/modules/services/networking/thelounge.nix:
```diff
@@ -0,0 +1,77 @@
+    };
+
+    port = mkOption {
+      type = types.int;
```
`types.port`

Infinisil commented on nixos/modules/services/networking/thelounge.nix:
```diff
@@ -0,0 +1,77 @@
+      script = "${pkgs.thelounge}/bin/thelounge start";
+      serviceConfig = {
+        DynamicUser = true;
+        StateDirectory = builtins.baseNameOf dataDir;
```
`baseNameOf` is in the global scope, so you can drop the `builtins.`

Infinisil commented on nixos/modules/services/networking/thelounge.nix:
```diff
@@ -0,0 +1,77 @@
+      default = false;
+      description = ''
+        Make your The Lounge instance private. You will need to configure user
+        accounts by using the provided wrapper script (`thelounge`) or by adding
```
`<command>thelounge</command>`

Infinisil commented on pkgs/top-level/all-packages.nix:
```diff
@@ -5859,6 +5859,8 @@ in
 
   theharvester = callPackage ../tools/security/theharvester { };
 
+  inherit(nodePackages) thelounge;
```
Nitpick: The conventional style is to have a space between `inherit` and `(..`

matthew-piziak commented:
Hi there! We're going to try a different workflow for `nodePackages`. Right now there are a number of PRs targeting `nodePackages`, and the churn makes it hard for you to keep your generated `.nix` files free of merge conflicts.

Could you please remove your generated changes, and just leave your changes in the `.json` file? That will make it easy to update, review, and merge. Then later this week we are going to generate everybody's `nodePackages` changes at once.

Thank you!

cc: @ryantm

Feel free to join us in Freenode's #nixos if you have questions about this.

Infinisil commented on nixos/modules/services/networking/thelounge.nix:
```diff
@@ -60,18 +53,26 @@ in {
-        DynamicUser = true;
-        StateDirectory = baseNameOf dataDir;
+        User = "thelounge";
+        StateDirectory = dataDir;
```
You still need `baseNameOf` here.

Infinisil commented on nixos/modules/services/networking/thelounge.nix:
```diff
@@ -60,18 +53,26 @@ in {
+      description = "thelounge service user";
+      group = "thelounge";
+      home = dataDir;
+      createHome = true;
```
No need for `home` and `createHome`, `StateDirectory` takes care of that.

Mrmaxmeier committed:
regenerate node-packages*.nix for nodePackages.thelounge

Mrmaxmeier commented:
@Infinisil I've addressed your review.
Is this ready to be merged?

Infinisil commented on nixos/modules/services/networking/thelounge.nix:
```diff
@@ -0,0 +1,76 @@
+      wantedBy = [ "multi-user.target" ];
+      environment = { THELOUNGE_HOME = dataDir; };
+      preStart = "ln -sf ${pkgs.writeText "config.js" configJsData} ${dataDir}/config.js";
+      script = "${pkgs.thelounge}/bin/thelounge start";
```
You can put this as `serviceConfig.ExecStart = "${pkgs.thelounge}/bin/thelounge start"`, which makes it have one less point of indirection.

Infinisil commented on nixos/modules/services/networking/thelounge.nix:
```diff
@@ -0,0 +1,76 @@
+    users.users.thelounge = {
+      description = "thelounge service user";
+      group = "thelounge";
+      isSystemUser = true;
```
Why does this need to be a system user? System users will have id's below 1000, but I see no reason why that would have to be the case.

Mrmaxmeier replied:
This is essentially a hint towards a service running as this user right?
I don't see any reason to assign specific UID ranges aside from a distinction between services / real users.

Is the size of the system user range a concern here?

Infinisil replied:
I haven't thought of it this way, but I've never seen a service use this. And the range is indeed pretty limited, from 400 to 499: https://github.com/NixOS/nixpkgs/blob/def7d562e01430ab48f4ab84a6363bfdbe6e584a/nixos/modules/programs/shadow.nix#L13-L14

So I think it's best to not do that unless the service needs a uid below 1000 for some reason

Mrmaxmeier replied:
It turns out most services on my machine are in the <400 uid range. I've dropped the system user requirement.

Infinisil replied:
> It turns out most services on my machine are in the <400 uid range

Ah yeah, so that's another issue, see https://github.com/NixOS/nixpkgs/issues/60732#issuecomment-488829636

Admittedly it is a bit odd to have services use uid's above 1000, which commonly are used by users, but that's the only way to get around the range restrictions for now until we have a better solution.

Infinisil replied:
Although actually, this 100 uid limit is only a problem if a user enables more than 100 `isSystemUser` services for a NixOS machine.

infinisil reviewed (changes requested):
Other than these little things this looks good to me. I can merge once those are addressed

infinisil reviewed (approved):
Alright this looks good to me :)

Now you just have to fix the merge conflict by updating the node set again (node packages are a bit annoying in nixpkgs..)

Feel free to ping me or @Mic92 when ready so we can merge

Infinisil commented:
Oh and one more thing: Would be nice if you could split up the commits into adding the package, updating node modules, then adding the NixOS module, in that order

Mrmaxmeier committed:
thelounge: init node package at v3.0.1

Mrmaxmeier committed:
nixos/thelounge: init

The Lounge is the official and community-managed fork of Shout.
This intends to replace the `shout` service.

Mrmaxmeier commented:
@Infinisil :ping_pong:

Infinisil commented:
NixOS module commits should start with "nixos/thelounge: ", and I'd put both last commits into a single one named "nixos/thelounge: init" (no version because the module is version-agnostic, and no separate commit for the non-dynamic thing because that was just addressing feedback from a review)

Mrmaxmeier commented:
Ok. I've rebased again and fixed up the commit log.

Infinisil commented:
Thanks for your patience :)
